src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+{
+    auto assetPair = std::make_pair(selling, buying);
+    auto iter = mLoadBestOfferContext.find(assetPair);
+    if (iter == mLoadBestOfferContext.end())
+    {
+        auto res = mLoadBestOfferContext.insert(std::make_pair(
+            assetPair, LoadBestOfferContext(selling, buying, *this)));
+        iter = res.first;
+    }
+    return iter->second;
+}
+
+void
+LedgerState::invalidateLoadBestOfferContext(Asset const& selling,
+                                            Asset const& buying)
+{
+    auto assetPair = std::make_pair(selling, buying);
+    auto iter = mLoadBestOfferContext.find(assetPair);
+    if (iter != mLoadBestOfferContext.end())
+    {
+        mLoadBestOfferContext.erase(iter);
+    }
+}
+
+std::vector<LedgerState::StateEntry>
+LedgerState::getOffers(Asset const& selling, Asset const& buying)
+{
+    std::vector<StateEntry> offers;
+    std::set<LedgerKey> seen;
+    getOffers(selling, buying, offers, seen);
+    return offers;
+}
+
+void
+LedgerState::getOffers(Asset const& selling, Asset const& buying,
+                       std::vector<StateEntry>& offers,
+                       std::set<LedgerKey>& seen)
+{
+    for (auto const& state : mState)
+    {
+        if (state.first.type() != OFFER)
+        {
+            continue;
+        }
+        else if (!seen.insert(state.first).second)
+        {
+            continue;
+        }
+        else if (!state.second->ignoreInvalid().entry())
+        {
+            continue;
+        }
+        assert(!state.second->valid());
+
+        // Note: Can't compare Assets with !=
+        auto const& offer = state.second->ignoreInvalid().entry()->data.offer();
+        if (!(selling == offer.selling) || !(buying == offer.buying))
+        {
+            continue;
+        }
+        offers.push_back(state.second);
+    }
+
+    if (mParent)
+    {
+        mParent->getOffers(selling, buying, offers, seen);
+    }
+}
+
+LedgerState::LoadBestOfferContext::LoadBestOfferContext(
+    Asset const& selling, Asset const& buying, LedgerState& ledgerState)
+    : mSelling(selling)
+    , mBuying(buying)
+    , mLedgerState(ledgerState)
+    , mTop(nullptr)
+    , mInMemory([](StateEntry const& lhs,
+                   StateEntry const& rhs) { return compareOffers(rhs, lhs); },
+                mLedgerState.getOffers(selling, buying))
+    , mOffersLoadedFromDatabase(0)
+{
+}
+
+void
+LedgerState::LoadBestOfferContext::loadFromDatabaseIfNecessary()
+{
+    if (mFromDatabase.empty())
+    {
+        LedgerState* root = &mLedgerState;
+        while (root->mParent != nullptr)
+        {
+            root = root->mParent;
+        }
+        mFromDatabase = root->loadBestOffersFromDatabase(
+            5, mOffersLoadedFromDatabase, mSelling, mBuying);
+        mOffersLoadedFromDatabase += mFromDatabase.size();
+    }
+}
+
+LedgerState::StateEntry
+LedgerState::LoadBestOfferContext::loadBestOffer()
+{
+    if (mTop)
```

Comment:
That inserts top value back into mInMemory. What is purpose of that? Won't it mean that each time the same value is returned?

Reply:
It does not mean that the same value is returned each time, although if the value is loaded and not modified or erased then it will be loaded again (as it should be, since it would still be the best offer at that point). The comment on line 686 gives an example where the offer loaded by `loadBestOffer` is erased, giving functionality similar to top followed by pop. Consider the case where the offer loaded by `loadBestOffer` is modified such that it is no longer the best offer. Then when it is pushed back into `mInMemory`, it will not necessarily be at the top of the heap.